### PR TITLE
cfmc30 proposal for FDCT

### PR DESCRIPTION
**Pull request by cfmc30**
---
A good PR should include the following parts:

- [x] Provide a clear title
- [x] Provide meaningful descriptions in the PR and the commit logs
- [x] Name your branch like ``<github-user-name>-hw#``, e.g., ``myname-hw2-attempt1``
- [x] Merge from your specific branch (***not master***) to the `hw*` branch (***not master***, again)
- [x] Exclude irrelevant files, such as binaries, text files, or dot files
- [x] Make sure tests/CI pass

### CODE REVIEW

File: project/cfmc30/README.rst [hunk 1]
```diff
@@ -0,0 +1,124 @@
+FDCT: Fast Discrete Cosine Transformation Library
+=================
+
+Basic Information
```
Comment (yungyuc) [inline project/cfmc30/README.rst:6]: Please supply the github repository for the project.

File: project/cfmc30/README.rst [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+===================
+
+FDCT is built as a Python extension using C++, and interacts with Python codebases
+through Numpy arrays as the primary data structure. The key architectural elements
+are:
+
+#. Input/Output
+
+   * FDCT accepts a 1D or 2D NumPy array as input, where each element represents a
+     pixel intensity (for images) or a signal value.
```
Comment (yungyuc) [inline project/cfmc30/README.rst]: Image uses 2D array and you will need to handle majoring. Which majoring you will support?  Please make it clear in your proposal.  A plan for the support should be included.

Please note that 2D array is required by image processing.  If you do not process image, you may or may not need to deal with 2D array.

File: project/cfmc30/README.rst [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+   The stable branch will merge when the code in dev branch is fully tested.
+3. Testing framework
+4. Documentation
+
+
+Schedule
+========
+
+* Planning phase (09/16-10/07)
+* Week 1 (10/14): Basic project infrastructure construction
```
Comment (yungyuc) [inline project/cfmc30/README.rst]: Add testing and CI to your development plan.  They should be available early in your development process and help you improve productivity.

Planning for the testing also help you modularize your system design.

File: project/cfmc30/README.rst [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+* Week 2 (10/21): Basic algorithm implementation
+* Week 3 (10/28): Code correctness validation
+* Week 4 (11/04): Optimization with SIMD and multithreading (1/2)
```
Comment (yungyuc) [inline project/cfmc30/README.rst:117]: SIMD and multithreading will take longer time than you plan.

File: project/cfmc30/README.rst [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+* Week 5 (11/11): Optimization with SIMD and multithreading (2/2)
+* Week 6 (11/18): Code correctness validation and benchmark
+* Week 7 (11/25): Binding with Python
+* Week 8 (12/02): Making Presentation slides and documentation
+
+References
```
Comment (yungyuc) [inline project/cfmc30/README.rst:123]: Very good to include references.  Good job.

**Review by yungyuc — CHANGES_REQUESTED** [review]
When requesting a next review, please add a global comment summarizing your change.

- [x] Add github repository link to your project.
- [ ] [L51](https://github.com/yungyuc/nsdhw_24au/pull/40/files#diff-81dffc67ecb2044a2b501bf4ae276a556092696cf91a6abee17e0750cbcf1addR51): Add plan for handling 2D array majoring.
- [x] [L108](https://github.com/yungyuc/nsdhw_24au/pull/40/files#diff-81dffc67ecb2044a2b501bf4ae276a556092696cf91a6abee17e0750cbcf1addR108): Add plan for testing and CI.

File: project/cfmc30/README.rst (hunk 2 above)
Comment (cfmc30) [reply project/cfmc30/README.rst]: OK, I plain to do both 1D and 2D array.

File: project/cfmc30/README.rst (hunk 4 above)
Comment (cfmc30) [reply project/cfmc30/README.rst:117]: OK, I will plain 3 weeks for SIMD and multithreading.

File: project/cfmc30/README.rst (hunk 3 above)
Comment (cfmc30) [reply project/cfmc30/README.rst]: CI added in planning phase.

**Comment by yungyuc** [comment]
> When requesting a next review, please add a global comment summarizing your change.

@cfmc30 Please follow the directive.

**Comment by cfmc30** [comment]
Following are changes:

- Add GitHub repository.
- Plain to do both 1D and 2D array.
- Plain 3 weeks for SIMD and multithreading.
- CI added in planning phase.

File: project/cfmc30/README.rst (hunk 1 above)
Comment (yungyuc) [reply project/cfmc30/README.rst:6]: Please do not resolve comments.  Resolving a comment makes it hard to track.  Leave them open.

File: project/cfmc30/README.rst (hunk 2 above)
Comment (yungyuc) [reply project/cfmc30/README.rst]: Please make it explicit (in the proposal) which majoring of the 2D arrays you will support.

File: project/cfmc30/README.rst (hunk 4 above)
Comment (yungyuc) [reply project/cfmc30/README.rst:117]: It usually takes months.  For now, it's OK to use any estimate in your plan.  Don't hesitate to change your course of development once you see the reality.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Do not resolve comments.  It makes tracking the discussions more difficult.

- [x] [L51](https://github.com/yungyuc/nsdhw_24au/pull/40/files#diff-81dffc67ecb2044a2b501bf4ae276a556092696cf91a6abee17e0750cbcf1addR51): Add plan for handling 2D array majoring.

**Comment by cfmc30** [comment]
Thanks for your suggestion and correction
Following are changes:

- Add GitHub repository.
- Focus on 2D array.
- Plain 3 weeks for SIMD and multithreading.
- CI added in planning phase.

File: project/cfmc30/README.rst [hunk 6]
```diff
@@ -0,0 +1,129 @@
+FDCT: Fast Discrete Cosine Transformation Library
+=================================================
+
+`GitHub Repository <https://github.com/cfmc30/FDCT>`_
+
+Basic Information
+=================
+
+FDCT is a Python library built using C++, designed to compute Discrete Cosine
+Transformations (DCT) and invert discrete cosine transformation, particularly
```
Comment (cfmc30) [inline project/cfmc30/README.rst:10]: Add focus on 2D array

File: project/cfmc30/README.rst (hunk 2 above)
Comment (cfmc30) [reply project/cfmc30/README.rst]: OK, I will explicitly emphasize the primary support for 2D arrays.

File: project/cfmc30/README.rst (hunk 4 above)
Comment (cfmc30) [reply project/cfmc30/README.rst:117]: Thanks for suggestion, I will try my best to following the schedule.

File: project/cfmc30/README.rst (hunk 1 above)
Comment (cfmc30) [reply project/cfmc30/README.rst:6]: Thanks for the clarification. I will make sure to leave the comments open moving forward.

File: project/cfmc30/README.rst (hunk 2 above)
Comment (yungyuc) [reply project/cfmc30/README.rst]: I still don't see your plan for [majoring](https://yyc.solvcon.net/en/latest/nsd/04matrix/matrix.html#row-major-2d-array).  Please take a look at the notes and research online and update the proposal.

File: project/cfmc30/README.rst (hunk 4 above)
Comment (yungyuc) [reply project/cfmc30/README.rst:117]: Just use reasonable efforts.  It is very easy to get absorbed into the detail optimization work and disregard everything else (like your research work and other coursework).  If you find it to be too time-consuming, do find me to discuss and change your project scope.

**Review by yungyuc — CHANGES_REQUESTED** [review]
You have not added any descriptions for majoring.  Please supplement.

**Comment by cfmc30** [comment]
Sorry for misunderstanding your comment, I have add the plan for majoring.
Changes:
- Row-major specified.

**Review by yungyuc — APPROVED** [review]
> Sorry for misunderstanding your comment, I have add the plan for majoring. Changes:
> 
> * Row-major specified.

LGTM.  For your information, column-major is popular for 2D arrays.  It's OK to focus on row-major arrays, but you could find both will be needed if you connect to other libraries.

Let me know when you want it to be merged.

**Pull request merged**